### PR TITLE
feat(sc): run the distillation teacher in the SingleController train pump

### PR DESCRIPTION
**Pull request by tianyi-zhang-02**
## What does this PR do?

Wires distillation into the SingleController train pump. It builds a frozen teacher on the training cluster, parks it between forwards, writes teacher top-k logits and indices through TransferQueue, fetches only the distillation loss fields, and skips reward/advantage work that distillation does not use.

The current version also keeps the newer mainline SC behavior intact: dynamic GRPO/PPO train fields, message-level penalties, token-capture schema, route fields, PPO critic epochs, and current setup validation.

This is the middle PR in #3843 → #3846 → #3849. The runnable recipe and checkpoint/restore functional test remain test-only in #3849.

## Ownership

The teacher top-k primitive overlaps #2580. I disclosed the duplicate there and in #3843; #2580 remains open and conflicting, with no author response since August 27. I am keeping this stack current and tested, but not extending the overlap further.

## Validation

- Head: `80966321d71eb204c42d3b51a322640c2f1964a6`
- Merged main: `e831622435180c88d6adcc9fce46bfd88a85f3e5`
- Linux x86_64, Python 3.13.14, CUDA disabled: **285 passed** across teacher top-k, SC distillation, entrypoint, setup, PPO setup, checkpointing, and controller actor suites.
- Ruff check and format: clean for all 13 changed Python files.
- The 2-GPU H100 checkpoint/restore run is recorded in #3849. It was run on the pre-refresh stack head `90760604`; the current main refresh was validated with CPU/config tests, not rerun on H100.

### CODE REVIEW

**Comment by tianyi-zhang-02** [comment]
**Heads-up on merge order, with the resolution written out.**

Four of my open PRs edit the same twelve-line block — the `unsupported` knob list in `_validate_algo_settings`:

- **#3786** removes `overlong_filtering` from it (`build_sample_mask` now applies it)
- **#3787** removes `reward_scaling` from it (`_advantage_stage` now applies it)
- **#3846 / #3849** wrap the list in an `is_distillation_run` guard, because `DistillationConfig` defines none of those knobs and the comprehension would `AttributeError`

I checked every pair by actually merging them. #3786 × #3787 is clean now — they only ever conflicted on the comment wording and on both adding a test before the same anchor, and both are fixed. The remaining four pairs are a genuine textual conflict on that block, and no restructuring avoids it: extracting or reordering the list conflicts just as hard.

So rather than force a stack, here is the resolution. Whichever order they land in, this is the merged form:

```python
    # An enabled one here describes shaping this run does not do. An entry
    # leaves this list when the SC path starts implementing it -- rejecting a
    # knob is only right while nobody honours it.
    #
    # DistillationConfig defines none of them: there is no reward to shape or
    # filter on, so the list cannot even be evaluated on that path.
    unsupported = (
        []
        if is_distillation_run(master_config)
        else [
            name
            for name, enabled in (
                ("use_dynamic_sampling", algo_cfg.use_dynamic_sampling),
                ("reward_scaling", algo_cfg.reward_scaling.enabled),
                ("reward_shaping", algo_cfg.reward_shaping.enabled),
            )
            if enabled
        ]
    )
```

I built that merge locally and ran the SC and config suites against it — 1084 passed. Happy to rebase whichever ones are left once the first lands; just say which order you want.

**Comment by tianyi-zhang-02** [comment]
Closing this dependent stack item together with #3843. The active distillation work now has separate ownership and direction in #2580 and the newer distillation series, so I do not want to maintain a competing SingleController integration.
